ocaml/lambda/translcore.ml

```diff
@@ -1177,24 +1204,78 @@ and transl_apply ~scopes
             }] in
           lfunction ~kind:(Curried {nlocal}) ~params
                     ~return:result_layout ~body ~mode ~region
-                    ~attr:default_stub_attribute ~loc
-        in
-        List.fold_right
-          (fun (id, layout, lam) body -> Llet(Strict, layout, id, lam, body))
-          !defs body
-    | Arg (arg, _) :: l -> build_apply lam (arg :: args) loc pos ap_mode l
+                    ~attr:default_stub_attribute ~loc)
+    | Arg (arg, _) :: l -> build_apply lam (arg :: args) loc pos ap_mode prot l
     | [] -> lapply lam (List.rev args) loc pos ap_mode result_layout
   in
-  let args =
-    List.map
-      (fun (_, arg) ->
-         match arg with
-         | Omitted _ as arg -> arg
-         | Arg (exp, sort_arg) ->
-           Arg (transl_exp ~scopes sort_arg exp, layout_exp sort_arg exp))
-      sargs
+
+  let rec build_apply_underscore lam rev_args loc prot = function
```

Comment:
change to "dummy" here

Reply:
This looks like a big block of code that's very similar to `build_apply` but with subtle differences. I'd really like to avoid that. (Also `prot` is just an optimization right? I don't think the old code has it. But I agree we'll run into a lot more so might as well avoid quadratic behavior.)

I realize why you've done this &mdash; you want to put the dummy parameters first in the lambda function &mdash; but can't we do this in a pre-processing step where we:

1. build up the entire arg list in the correct order
2. reorder the original arg list to put the dummy arguments first
3. call `build_apply` where the eventual body is an application to the correctly-ordered arg list (not the reordered arg list)

There might be some reason that this doesn't work, but if it does work, I think we can avoid duplicating the logic in `build_apply` and here.

Reply:
I feel that `Omitted` and `Dummy` are different enough that it's easier to process them in two passes (like what we have now). But I've mechanically factor out the common part of the two functions to reduce duplication.

Reply:
We might take a different approach in light of my most recent comment (the function we build shouldn't be a series of nested unary functions for dummy arguments) -- perhaps the code sharing will be dramatically less and we won't need `lfunction_helper`.